File: karavan-web/karavan-app/src/main/java/org/apache/camel/karavan/service/NotificationService.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.camel.karavan.service;

import io.vertx.core.eventbus.DeliveryOptions;
import io.vertx.core.json.JsonObject;
import io.vertx.mutiny.core.eventbus.EventBus;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;

import java.util.UUID;

import static org.apache.camel.karavan.shared.Constants.*;

@ApplicationScoped
public class NotificationService {

    @Inject
    EventBus eventBus;

    public void send(String userId, String eventId, String evenName, String className, JsonObject data) {
        eventBus.publish(userId, data, new DeliveryOptions()
                        .addHeader(NOTIFICATION_HEADER_EVENT_ID, eventId != null ? eventId : UUID.randomUUID().toString())
                        .addHeader(NOTIFICATION_HEADER_EVENT_NAME, evenName)
                        .addHeader(NOTIFICATION_HEADER_CLASS_NAME, className)
        );
    }

    public void sendSystem(String eventId, String evenName, String className, JsonObject data) {
        eventBus.publish(NOTIFICATION_ADDRESS_SYSTEM, data, new DeliveryOptions()
                        .addHeader(NOTIFICATION_HEADER_EVENT_ID, eventId != null ? eventId : UUID.randomUUID().toString())
                        .addHeader(NOTIFICATION_HEADER_EVENT_NAME, evenName)
                        .addHeader(NOTIFICATION_HEADER_CLASS_NAME, className)
        );
    }
}

File: karavan-app/src/main/webui/src/api/NotificationService.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {Subject} from "rxjs";
import {unstable_batchedUpdates} from "react-dom";
import {useProjectStore} from "./ProjectStore";
import {ProjectService} from "./ProjectService";
import {EventBus} from "../designer/utils/EventBus";

export class KaravanEvent {
    id: string = '';
    type: 'system' | 'user' = 'system';
    event: string = '';
    className: string = '';
    data: any = {};

    public constructor(init?: Partial<KaravanEvent>) {
        Object.assign(this, init);
    }
}

const karavanEvents = new Subject<KaravanEvent>();

export const NotificationEventBus = {
    sendEvent: (event: KaravanEvent) =>  karavanEvents.next(event),
    onEvent: () => karavanEvents.asObservable(),
}

const sub = NotificationEventBus.onEvent()?.subscribe((event: KaravanEvent) => {
    if (event.event === 'configShared') {
        const filename = event.data?.filename ? event.data?.filename : 'all'
        EventBus.sendAlert('Success', 'Configuration shared for ' + filename);
    } else if (event.event === 'commit' && event.className === 'Project') {
        const projectId = event.data?.projectId;
        if (useProjectStore.getState().project?.projectId === projectId) {
            unstable_batchedUpdates(() => {
                useProjectStore.setState({isPushing: false});
                ProjectService.refreshProject(projectId);
                ProjectService.refreshProjectData(projectId);
            });
        }
    } else if (event.event === 'imagesLoaded') {
        const projectId = event.data?.projectId;
        unstable_batchedUpdates(() => {
            ProjectService.refreshImages(projectId);
        });
    } else if (event.event === 'error') {
        const error = event.data?.error;
        EventBus.sendAlert('Error', error, "danger");
    } else if (event.event === 'ping') {
        // do nothing
    } else {
        const message = event.data?.message ?  event.data?.message : JSON.stringify(event.data);
        EventBus.sendAlert('Success', message);
    }
});

